File: app/src/main/java/com/chebuso/chargetimer/repositories/CalendarEntityReader.java
package com.chebuso.chargetimer.repositories;

import android.database.Cursor;
import android.provider.CalendarContract;

import com.chebuso.chargetimer.models.CalendarEntity;

class CalendarEntityReader {
    CalendarEntity fromCursorPosition(Cursor cur){
        int idIndex = cur.getColumnIndex(CalendarContract.Calendars._ID);
        int accountNameIndex = cur.getColumnIndex(CalendarContract.Calendars.ACCOUNT_NAME);
        int accountTypeIndex = cur.getColumnIndex(CalendarContract.Calendars.ACCOUNT_TYPE);
        int ownerAccountIndex = cur.getColumnIndex(CalendarContract.Calendars.OWNER_ACCOUNT);
        int displayViewIndex = cur.getColumnIndex(CalendarContract.Calendars.CALENDAR_DISPLAY_NAME);
        int visibleIndex = cur.getColumnIndex(CalendarContract.Calendars.VISIBLE);
        int isPrimaryIndex = cur.getColumnIndex(CalendarContract.Calendars.IS_PRIMARY);
        if (isPrimaryIndex == -1) {
            isPrimaryIndex = cur.getColumnIndex("COALESCE(isPrimary, ownerAccount = account_name)");
        }

        CalendarEntity result = new CalendarEntity();

        result.id = cur.getLong(idIndex);
        result.accountName = cur.getString(accountNameIndex);
        result.accountType = cur.getString(accountTypeIndex);
        result.ownerAccount = cur.getString(ownerAccountIndex);
        result.displayName = cur.getString(displayViewIndex);
        result.visible = cur.getInt(visibleIndex) == 1;

        int isPrimary = isPrimaryIndex== -1 ? -1 : cur.getInt(isPrimaryIndex);
        if (isPrimary == -1 && result.accountName.equals(result.ownerAccount))
            isPrimary = 1;

        result.isPrimary = isPrimary == 1;
        return result;
    }
}

File: EvChargeTimer/app/src/main/java/com/chebuso/chargetimer/calendar/dal/CalendarEntityReader.kt
package com.chebuso.chargetimer.calendar.dal

import android.database.Cursor
import android.provider.CalendarContract
import com.chebuso.chargetimer.calendar.CalendarEntity

internal class CalendarEntityReader {
    fun fromCursorPosition(cur: Cursor): CalendarEntity {
        val idIndex = cur.getColumnIndex(CalendarContract.Calendars._ID)
        val accountNameIndex = cur.getColumnIndex(CalendarContract.Calendars.ACCOUNT_NAME)
        val accountTypeIndex = cur.getColumnIndex(CalendarContract.Calendars.ACCOUNT_TYPE)
        val ownerAccountIndex = cur.getColumnIndex(CalendarContract.Calendars.OWNER_ACCOUNT)
        val displayViewIndex = cur.getColumnIndex(CalendarContract.Calendars.CALENDAR_DISPLAY_NAME)
        val visibleIndex = cur.getColumnIndex(CalendarContract.Calendars.VISIBLE)
        val isPrimaryIndex = getIsPrimaryIndex(cur)

        return CalendarEntity().apply {
            id = cur.getLong(idIndex)
            accountName = cur.getString(accountNameIndex)
            accountType = cur.getString(accountTypeIndex)
            ownerAccount = cur.getString(ownerAccountIndex)
            displayName = cur.getString(displayViewIndex)
            visible = cur.getInt(visibleIndex) == 1
            isPrimary = getIsPrimary(isPrimaryIndex, cur)
        }
    }

    private fun getIsPrimaryIndex(cur: Cursor): Int {
        val isPrimaryIndex = cur.getColumnIndex(CalendarContract.Calendars.IS_PRIMARY)
        return if (isPrimaryIndex == -1)
            cur.getColumnIndex("COALESCE(isPrimary, ownerAccount = account_name)")
        else
            isPrimaryIndex
    }

    private fun getIsPrimary(isPrimaryIndex: Int, cur: Cursor): Boolean {
        val isPrimary = if (isPrimaryIndex == -1) 0 else cur.getInt(isPrimaryIndex)
        return isPrimary == 1
    }
}

